### PR TITLE
feat(status): show validator heartbeats per period

### LINKED ISSUE #259: heartbeat display in bridge-ui
## Problem Description

- when a prospect validator submits the [registration request](https://docs.google.com/forms/d/e/1FAIpQLSdQtc5LoEyWkc5-86SOLW3xK8cRNwuByC7SIrA9MdWeAiuBZw/viewform) the testnet/mainnet commander will mint a heartbeat token, and send it to the address of the validator
- once the heartbeat token arrived, the validator will start spending it regularly
- the commander needs to be able to measure the uptime over a period of x hours/days, to be sure that the validator is set up well, before assigning a slot.

## Scope

- create a screen in bridge-ui that lists all slots (see git history, once existed)
- calculate and display uptime of slot based on heartbeat token spending

## Deliverables

- code
- tests

## Gain for the project

- unblock Ops circle, to unblock ECO system for validator launch

## Expert / Affected

affected: @troggy 
expert: @johannbarbie 

## Roles
bounty gardener: @simonweniger  / 10%
bounty worker: name / share
bounty reviewer: name / share

### PR DESCRIPTION
**Pull request by troggy**
Please remember the [Contributing Guidelines](https://github.com/leapdao/meta/blob/master/CONTRIBUTION.md) :heart:

## What does this PR do?

For each of the validator slot, get the history of the heartbeat tx and show which periods are "heartbeated" and which are not.

image

## Recommendations for testing

1. Run local network (`yarn start` in leap-sandbox)
2. Run bridge ui for local network (`yarn start`)
3. Push local network to the next period. If you push fast enough heartbeat will be missed.

Alternatively
1. Run root local network (`yarn start --onlyRoot` in leap-sandbox)
2. Setup a validator (use `setValidator` from leap-guardian)
3. Push few periods ("heartbeated")
4. stop the node
5. disable heartbeat in the network config (e.g. remove the `heartbeat.color`)
6. run the node again
7. push few periods (not heartbeated)
8. etc mix and improvise

## Links to relevant issues or information

Closes #259

### CODE REVIEW

**Comment by troggy** [comment]
@sunify @MaxStalker need ruthless review for this (probably not great) mobx-lasagna. 100 DAI reviewer share

**Comment by MaxStalker** [comment]
@troggy I gave it a quick glance today. On the surface - looks quite good :)
Will give more in-depth check in the morning.

File: src/components/heartbeats/heartbeatsStore.ts [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+      return;
+    }
+    return this.loadHeartbeatPeriodsRecursive(
+      slotHeartbeatsData,
+      tx.inputs[0].prevout.hash
+    );
+  }
+
+  private async loadHeartbeatsForSlot(slotId) {
+    const [heartbeatUtxo] = await web3PlasmaStore.instance.getUnspent(
```
Comment (MaxStalker) [inline src/components/heartbeats/heartbeatsStore.ts]: I thought we've disabled this for heartbeat color... 🤔

File: src/components/heartbeats/heartbeatsStore.ts [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+  constructor(keepPeriodsLimit = 15) {
+    this.keepPeriodsLimit = keepPeriodsLimit;
+
+    reaction(
+      () => operatorStore.slots,
+      () => {
+        this.storageKey = `hb-${operatorStore.address}`;
+        this.heartbeats = this.initHeartbeatsData();
+        operatorStore.contract.methods
+          .heartbeatColor()
```
Comment (MaxStalker) [inline src/components/heartbeats/heartbeatsStore.ts]: Currently this method is not supported on testnet (needs an update of poaContract, I guess).

File: src/components/heartbeats/heartbeatsStore.ts (hunk 1 above)
Comment (troggy) [reply src/components/heartbeats/heartbeatsStore.ts]: that's a great catch! Totally overlooked — leap-sandbox runs a node with filtering switched off.

Comment (troggy) [reply src/components/heartbeats/heartbeatsStore.ts]: I will fix this. Thanks for the catch

Comment (troggy) [reply src/components/heartbeats/heartbeatsStore.ts]: addressed in latest commit. The only idea I can come up is to scan blocks a-la dapparatus/burner-wallet. Btw, it is so easy to bypass the filtering and push your heartbeat even without running the node at all.

Comment (MaxStalker) [reply src/components/heartbeats/heartbeatsStore.ts]: Yap, that what I discovered as well - sorry if I forgot to mention it 🙇‍♂️
I was thinking to wait for new period, then some more blocks after it and then parse them and find those that look like heartbeat. Assuming that validator will make heartbeat tx in first couple of blocks 🤷‍♂️

**Pull request merged**